vite-template-redux/src/app/effect.ts

```diff
@@ -0,0 +1,107 @@
+import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
+import type { BaseQueryApi } from '@reduxjs/toolkit/query';
+import type { Context } from 'effect';
+import { Effect, Either } from 'effect';
+import type { ThunkExtraArgument } from '../app/store';
+
+// Helper type that extracts the success type from an Effect
+type ExtractSuccessType<T> = T extends Effect.Effect<infer A, any, any> ? A : never;
+
+// Helper type that extracts method names (as strings) from a service interface
+type MethodKeys<T> = {
+  [K in keyof T]: T[K] extends (...args: any[]) => any ? (K extends string ? K : never) : never;
+}[keyof T];
+
+// Helper type that extracts all possible error types from service methods
+type ExtractErrorTypes<S> = {
+  [K in keyof S]: S[K] extends (...args: any[]) => Effect.Effect<any, infer E, any> ? E : never;
+}[keyof S];
+
+// Main function that creates an RTK Query API from an Effect Layer
+export const createApiFromEffectLayer = <
+  S, 
+  SI extends { [K in keyof SI]: (...args: any[]) => Effect.Effect<any, any, any> }
+>(
+  // Takes a service tag (Context.Tag) and configuration options
+  serviceTag: Context.Tag<S, SI>,
+  options: {
+    reducerPath: string;
+  } & {
+    // For each method, specify if it's a query or mutation and optional cache tags
+    [K in MethodKeys<SI>]: {
+      type: 'query' | 'mutation';
+      providesTags?: string[];
+      invalidatesTags?: string[];
+    };
+  } 
+) => {
+  type ErrorType = ExtractErrorTypes<SI>;
+
+  // Helper function to create a queryFn for RTK Query endpoints
+  const createQueryFn = (methodKey: keyof SI) => {
+    // method is used for typing purposes, not for runtime
+    const method = (serviceTag as any)[methodKey] as SI[typeof methodKey];
+
+    return async (args: Parameters<typeof method>[0], api: BaseQueryApi) => {
+      const effect = Effect.gen(function* () {
+        const service = yield* serviceTag;
+        return yield* service[methodKey](args);
+      }).pipe(Effect.either);
```

Comment:
We've got some use cases in which we need to access redux state in the query handlers (for example to get JWT token for currently logged in user), so would be good to do something like:

```ts
export class ReduxState extends Context.Tag("ReduxState")<
  ReduxState,
  unknown
>() {}
```

and then

```suggestion
      const effect = Effect.gen(function* () {
        const service = yield* serviceTag;
        return yield* service[methodKey](args);
      }).pipe(
        Effect.provideService(ReduxState, api.getState()), 
        Effect.either
      );
```

We can leave state typed as `unknown` (since that's what RTK does) and leave it to the code in the effect to cast or parse it as necessary.

Reply:
Yup, makes sense. I can see other use cases for injecting app state into effects.